dev/com.ibm.ws.classloading/src/com/ibm/ws/classloading/internal/ContainerClassLoader.java

```diff
@@ -1080,64 +1090,89 @@ private static class SmartClassPathImpl implements SmartClassPath {
                     @Override
                     public void run() {
-                        if (tc.isDebugEnabled()) {
-                            if (uc instanceof ArtifactContainerUniversalContainer) {
-                                Tr.debug(tc, "CCL: " + this.hashCode() + " building package map for " + ((ArtifactContainerUniversalContainer) uc).container.getPhysicalPath());
-                            } else {
-                                Tr.debug(tc, "CCL: " + this.hashCode() + " building package map for " + ((ContainerUniversalContainer) uc).container.getPhysicalPath());
-                            }
-                        }
-                        //perform the update of the map inside the write lock
-                        //to prevent the classloader using the map in an inconsistent state.
-                        WriteLock write = rwLock.writeLock();
-                        write.lock();
                         try {
-                            uc.updatePackageMap(packageMap);
-                            outstandingContainers.decrementAndGet();
+                            for (UniversalContainer uc : ucs) {
+                                if (tc.isDebugEnabled()) {
+                                    if (uc instanceof ArtifactContainerUniversalContainer) {
+                                        Tr.debug(tc, "CCL: " + this.hashCode() + " building package map for " + ((ArtifactContainerUniversalContainer) uc).container.getPhysicalPath());
+                                    } else {
+                                        Tr.debug(tc, "CCL: " + this.hashCode() + " building package map for " + ((ContainerUniversalContainer) uc).container.getPhysicalPath());
+                                    }
+                                }
+                                //perform the update of the map inside the write lock
+                                //to prevent the classloader using the map in an inconsistent state.
+                                WriteLock write = rwLock.writeLock();
+                                write.lock();
+                                try {
+                                    uc.updatePackageMap(packageMap);
+                                } finally {
+                                    write.unlock();
+                                }
+                                if (tc.isDebugEnabled())
+                                    Tr.debug(tc, "CCL: " + this.hashCode() + " done building package map.");
+                            }
                         } finally {
-                            write.unlock();
+                            outstandingContainers.decrementAndGet();
                         }
-                        if (tc.isDebugEnabled())
-                            Tr.debug(tc, "CCL: " + this.hashCode() + " done building package map.");
                     }
-                });
-
+                };
+                if (synchronous) {
```

Comment:
Instead of having to pass down a new `synchronous` variable that rips through the internal methods I suggest checking for `!CheckpointPoint.getPhase().restored()` here.

If you are worried about a call to the synchronized `getPhase()` method the `CheckpointPhase` can safely be stored in a `static final` variable for the this class.